File: services/feature-dispatch/src/runtimeEnvironment.js
import { readFileSync } from "node:fs";

export function environmentValue(name, {
  environment = process.env,
  required = true,
} = {}) {
  let value = String(environment[name] || "").trim();
  const filePath = String(environment[`${name}_FILE`] || "").trim();
  if (value && filePath) throw new Error(`${name} and ${name}_FILE cannot both be set`);
  if (filePath) {
    try {
      value = readFileSync(filePath, "utf8").trim();
    } catch (error) {
      throw new Error(`cannot read ${name}_FILE`, { cause: error });
    }
  }
  if (required && !value) throw new Error(`${name} or ${name}_FILE is required`);
  return value || null;
}

File: services/feature-engine/src/feature_engine/runtime_environment.py
from __future__ import annotations

from pathlib import Path
from typing import Mapping


def environment_value(
    environment: Mapping[str, str],
    name: str,
    *,
    required: bool = True,
) -> str | None:
    direct = str(environment.get(name) or "").strip()
    file_path = str(environment.get(f"{name}_FILE") or "").strip()
    if direct and file_path:
        raise RuntimeError(f"{name} and {name}_FILE cannot both be set")
    if file_path:
        try:
            direct = Path(file_path).read_text(encoding="utf-8").strip()
        except OSError as error:
            raise RuntimeError(f"cannot read {name}_FILE") from error
    if required and not direct:
        raise RuntimeError(f"{name} or {name}_FILE is required")
    return direct or None


def required_environment(environment: Mapping[str, str], name: str) -> str:
    value = environment_value(environment, name)
    assert value is not None
    return value


def optional_environment(environment: Mapping[str, str], name: str) -> str | None:
    return environment_value(environment, name, required=False)
